scalafix-core/src/main/scala/scalafix/Scalafix.scala

```diff
@@ -1,50 +1,23 @@
 package scalafix
 
-import scala.collection.immutable.Seq
 import scala.meta._
-import scala.meta.inputs.Input
-import scala.meta.parsers.Parsed
-import scalafix.config.ScalafixConfig
-import scalafix.rewrite.RewriteCtx
-import scalafix.rewrite.ScalafixCtx
-import scalafix.rewrite.ScalafixMirror
-import scalafix.util.CanOrganizeImports
-import scalafix.util.Patch
+import scala.util.control.NonFatal
 
 object Scalafix {
-  def fix(code: Input,
-          config: ScalafixConfig,
-          semanticApi: Option[ScalafixMirror]): Fixed = {
-    config.parser.apply(code, config.dialect) match {
-      case Parsed.Success(ast) =>
-        fix(ast, config, semanticApi)
-      case Parsed.Error(pos, msg, e) =>
-        Fixed.Failed(Failure.ParseError(pos, msg, e))
+  def fix(input: Input, config: ScalafixConfig): Fixed =
+    config.dialect(input).parse[Source] match {
+      case parsers.Parsed.Success(ast) =>
+        fix(RewriteCtx.syntactic(ast, config))
+      case parsers.Parsed.Error(pos, msg, details) =>
+        Fixed.Failed(Failure.ParseError(pos, msg, details))
     }
-  }
-  def fix(ast: Tree,
-          config: ScalafixConfig,
-          semanticApi: Option[ScalafixMirror]): Fixed = {
-    implicit val ctx: ScalafixCtx = RewriteCtx(
-      ast,
-      config,
-      semanticApi.getOrElse(ScalafixMirror.failingMirror(config.dialect)))
+
+  def fix(ctx: RewriteCtx): Fixed =
     try {
-      val patches = config.rewrites.flatMap(_.rewrite(ctx))
-      val patched =
-        Patch.apply(patches)(CanOrganizeImports.ScalafixMirror, ctx)
-      Fixed.Success(patched)
+      Fixed.Success(ctx.config.rewrite.wrappedRewrite(ctx).applied)
     } catch {
-      case e: Failure =>
-        Fixed.Failed(e)
+      case NonFatal(e) =>
```

Comment:
I don't like this at all. However, there are quite many things that can go wrong here since we are calling user-implemented Rewrites and I don't want to crash the scalafix-cli when fixing hundreds of files only because one file failed.